File: backend/app/models/template.py
"""
Template Model
Defines the structure for photo booth templates (layout + design)

A Template combines:
- Photo layout (how many photos and their arrangement)
- Design assets (background, logos from Canva)
- Positioning rules (where design elements go)
"""
from typing import Literal
from pydantic import BaseModel, Field
from datetime import datetime


# Constants - Avoid magic strings
LAYOUT_VERTICAL_3 = "3x1-vertical"
LAYOUT_VERTICAL_4 = "4x1-vertical"
LAYOUT_VERTICAL_6 = "6x1-vertical"
LAYOUT_GRID_2X2 = "2x2-grid"

DESIGN_POSITION_TOP = "top"
DESIGN_POSITION_BOTTOM = "bottom"
DESIGN_POSITION_LEFT = "left"
DESIGN_POSITION_RIGHT = "right"

# Valid layout options
LayoutType = Literal["3x1-vertical", "4x1-vertical", "6x1-vertical", "2x2-grid"]
DesignPositionType = Literal["top", "bottom", "left", "right"]


class Template(BaseModel):
    """
    Template defines a complete photo strip layout.
    
    Separates concerns:
    - Layout: How photos are arranged
    - Design: Visual assets (logos, backgrounds)
    - Settings: Spacing, colors, etc.
    """
    id: str
    name: str = Field(description="User-friendly template name")
    
    # Layout configuration
    layout: LayoutType = Field(
        default=LAYOUT_VERTICAL_3,
        description="Photo arrangement pattern"
    )
    
    # Design assets
    design_file_path: str | None = Field(
        default=None,
        description="Path to design PNG from Canva"
    )
    design_position: DesignPositionType = Field(
        default=DESIGN_POSITION_BOTTOM,
        description="Where to place the design element"
    )
    
    # Visual settings
    background_color: str = Field(
        default="#FFFFFF",
        description="Hex color code for background"
    )
    photo_spacing: int = Field(
        default=20,
        ge=0,
        le=100,
        description="Spacing between photos in pixels"
    )
    
    # Metadata
    is_active: bool = Field(
        default=False,
        description="Whether this template is currently active"
    )
    created_at: str = Field(
        default_factory=lambda: datetime.now().isoformat(),
        description="ISO timestamp of creation"
    )
    
    # Optional preview
    preview_url: str | None = Field(
        default=None,
        description="URL to preview image"
    )


class TemplateCreate(BaseModel):
    """Request model for creating a new template"""
    name: str = Field(min_length=1, max_length=100)
    layout: LayoutType = LAYOUT_VERTICAL_3
    design_position: DesignPositionType = DESIGN_POSITION_BOTTOM
    background_color: str = "#FFFFFF"
    photo_spacing: int = Field(default=20, ge=0, le=100)


class TemplateUpdate(BaseModel):
    """Request model for updating a template - all fields optional"""
    name: str | None = None
    layout: LayoutType | None = None
    design_file_path: str | None = None
    design_position: DesignPositionType | None = None
    background_color: str | None = None
    photo_spacing: int | None = Field(default=None, ge=0, le=100)


def get_layout_photo_count(layout: LayoutType) -> int:
    """
    Returns the number of photos for a given layout.
    Single source of truth for layout -> photo count mapping.
    
    Args:
        layout: The layout type
        
    Returns:
        Number of photos in this layout
    """
    layout_map = {
        LAYOUT_VERTICAL_3: 3,
        LAYOUT_VERTICAL_4: 4,
        LAYOUT_VERTICAL_6: 6,
        LAYOUT_GRID_2X2: 4,
    }
    return layout_map[layout]


def get_layout_dimensions(layout: LayoutType) -> tuple[int, int]:
    """
    Returns (width, height) in pixels for a given layout.
    
    Args:
        layout: The layout type
        
    Returns:
        Tuple of (width, height) in pixels
    """
    # Standard strip width
    strip_width = 600
    
    # Heights based on number of photos + design space
    dimension_map = {
        LAYOUT_VERTICAL_3: (strip_width, 1800),  # 3 photos + design
        LAYOUT_VERTICAL_4: (strip_width, 2400),  # 4 photos + design
        LAYOUT_VERTICAL_6: (strip_width, 3000),  # 6 photos + design
        LAYOUT_GRID_2X2: (strip_width, 1800),    # 2x2 grid + design
    }
    return dimension_map[layout]

File: frontend-new/src/types/template.ts
/**
 * Template Types
 * Defines TypeScript types for photo booth templates
 * 
 * A Template = Layout + Design + Positioning
 */

// Constants - Avoid magic strings
export const LAYOUT_3X1_VERTICAL = '3x1-vertical' as const;
export const LAYOUT_4X1_VERTICAL = '4x1-vertical' as const;
export const LAYOUT_6X1_VERTICAL = '6x1-vertical' as const;
export const LAYOUT_2X2_GRID = '2x2-grid' as const;

export const DESIGN_POSITION_TOP = 'top' as const;
export const DESIGN_POSITION_BOTTOM = 'bottom' as const;
export const DESIGN_POSITION_LEFT = 'left' as const;
export const DESIGN_POSITION_RIGHT = 'right' as const;

export const OVERLAY_MODE_FREE = 'free' as const;
export const OVERLAY_MODE_FOOTER = 'footer' as const;

// Type definitions
export type LayoutType =
  | typeof LAYOUT_3X1_VERTICAL
  | typeof LAYOUT_4X1_VERTICAL
  | typeof LAYOUT_6X1_VERTICAL
  | typeof LAYOUT_2X2_GRID;

export type DesignPositionType =
  | typeof DESIGN_POSITION_TOP
  | typeof DESIGN_POSITION_BOTTOM
  | typeof DESIGN_POSITION_LEFT
  | typeof DESIGN_POSITION_RIGHT;

export type OverlayModeType =
  | typeof OVERLAY_MODE_FREE
  | typeof OVERLAY_MODE_FOOTER;

export type PhotoAspectRatioType = 'auto' | '1:1' | '3:4';

/**
 * Template model - matches backend exactly
 */
export interface Template {
  id: string;
  name: string;
  layout: LayoutType;
  design_file_path: string | null;
  design_position: DesignPositionType;
  overlay_mode: OverlayModeType;
  // Optional free overlay controls (normalized / scale)
  // When all are null/undefined, backend usa el modo legacy de banda fija.
  design_scale?: number | null;
  design_offset_x?: number | null;
  design_offset_y?: number | null;
  design_stretch?: boolean;
  background_color: string;
  photo_spacing: number;
  photo_filter?: 'none' | 'bw' | 'sepia' | 'glam';
  photo_aspect_ratio?: PhotoAspectRatioType;
  is_active: boolean;
  created_at: string;
  preview_url: string | null;
}

/**
 * Request model for creating a template
 */
export interface TemplateCreate {
  name: string;
  layout: LayoutType;
  design_position: DesignPositionType;
  overlay_mode: OverlayModeType;
  background_color: string;
  photo_spacing: number;
  photo_filter?: 'none' | 'bw' | 'sepia' | 'glam';
  design_scale?: number | null;
  design_offset_x?: number | null;
  design_offset_y?: number | null;
  design_stretch?: boolean;
  photo_aspect_ratio?: PhotoAspectRatioType;
}

/**
 * Request model for updating a template
 */
export interface TemplateUpdate {
  name?: string;
  layout?: LayoutType;
  design_file_path?: string;
  design_position?: DesignPositionType;
  overlay_mode?: OverlayModeType;
  background_color?: string;
  photo_spacing?: number;
  photo_filter?: 'none' | 'bw' | 'sepia' | 'glam';
  design_scale?: number | null;
  design_offset_x?: number | null;
  design_offset_y?: number | null;
  design_stretch?: boolean;
  photo_aspect_ratio?: PhotoAspectRatioType;
}

export interface FormData {
  name: string;
  layout: LayoutType;
  design_file: File | null;
  design_position: DesignPositionType;
  overlay_mode: OverlayModeType;
  background_color: string;
  photo_spacing: number;
  photo_filter: 'none' | 'bw' | 'sepia' | 'glam';
  design_scale: number;
  design_offset_x: number;
  design_offset_y: number;
  design_stretch: boolean;
  photo_aspect_ratio: PhotoAspectRatioType;
}

/**
 * Response model for templates list
 */
export interface TemplatesListResponse {
  templates: Template[];
  active_template: Template | null;
}

/**
 * Response model for design upload
 */
export interface UploadDesignResponse {
  success: boolean;
  file_path: string;
  message: string;
}

/**
 * Human-readable labels for layouts
 */
export const LAYOUT_LABELS: Record<LayoutType, string> = {
  [LAYOUT_3X1_VERTICAL]: '3 fotos verticales',
  [LAYOUT_4X1_VERTICAL]: '4 fotos verticales',
  [LAYOUT_6X1_VERTICAL]: '6 fotos verticales',
  [LAYOUT_2X2_GRID]: 'Grid 2x2',
};

/**
 * Human-readable labels for design positions
 */
export const DESIGN_POSITION_LABELS: Record<DesignPositionType, string> = {
  [DESIGN_POSITION_TOP]: 'Arriba',
  [DESIGN_POSITION_BOTTOM]: 'Abajo',
  [DESIGN_POSITION_LEFT]: 'Izquierda',
  [DESIGN_POSITION_RIGHT]: 'Derecha',
};

/**
 * Helper: Get photo count for a layout
 * Single source of truth for layout -> count mapping
 */
export function getLayoutPhotoCount(layout: LayoutType): number {
  const countMap: Record<LayoutType, number> = {
    [LAYOUT_3X1_VERTICAL]: 3,
    [LAYOUT_4X1_VERTICAL]: 4,
    [LAYOUT_6X1_VERTICAL]: 6,
    [LAYOUT_2X2_GRID]: 4,
  };
  return countMap[layout];
}

/**
 * Helper: Get layout dimensions in pixels
 */
export function getLayoutDimensions(layout: LayoutType): { width: number; height: number } {
  const stripWidth = 600;

  const dimensionMap: Record<LayoutType, { width: number; height: number }> = {
    [LAYOUT_3X1_VERTICAL]: { width: stripWidth, height: 1800 },
    [LAYOUT_4X1_VERTICAL]: { width: stripWidth, height: 2400 },
    [LAYOUT_6X1_VERTICAL]: { width: stripWidth, height: 3000 },
    [LAYOUT_2X2_GRID]: { width: stripWidth, height: 1800 },
  };

  return dimensionMap[layout];
}
